File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int maxLength = 0;
        int i = 0,j = 0;
        unordered_set<char> set;

        while (i < n && j < n) {
            if (set.find(s[j]) == set.end()) {
                set.insert(s[j++]);
                maxLength = max(maxLength, j - i);
            } 
            else {
                set.erase(s[i++]);
            }
        }

        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLength = 0;
        int left = 0;
        HashMap<Character, Integer> charIndexMap = new HashMap<>();

        for (int right = 0; right < n; right++) {
            char ch = s.charAt(right);

            if (charIndexMap.containsKey(ch)) {
                left = Math.max(left, charIndexMap.get(ch) + 1);
            }

            charIndexMap.put(ch, right);
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}